providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+        goto fail;
+
+    if (EVP_MAC_init(ctx) != 1)
+        goto fail;
+
+    if (EVP_MAC_update(ctx, in, inlen) != 1)
+        goto fail;
+
+    if (EVP_MAC_final(ctx, out, (size_t *) &out_written, outlen) != 1)
+        goto fail;
+
+    ret = 1;
+
+fail:
+    EVP_MAC_free_ctx(ctx);
+    EVP_MAC_free(mac);
+    return ret;
+}
+
+static int blake2b_long(unsigned char *out, size_t outlen, const void *in,
+                        size_t inlen)
+{
+    int ret;
+    EVP_MAC *mac;
+    EVP_MAC_CTX *ctx;
+    OSSL_PARAM par[2];
+    size_t par_n, out_written, blake_outlen;
+    uint32_t outlen_curr;
+    uint8_t outbuf[BLAKE2B_OUTBYTES];
+    uint8_t inbuf[BLAKE2B_OUTBYTES];
+    uint8_t outlen_bytes[sizeof(uint32_t)] = {0};
+
+    ctx = NULL;
+    mac = NULL;
+    ret = 0;
+    par_n = 0;
+    blake_outlen = outlen;
+
+    if (outlen > UINT32_MAX || out == NULL || outlen == 0)
+        goto fail;
+
+    if (outlen > BLAKE2B_OUTBYTES)
+        blake_outlen = BLAKE2B_OUTBYTES;
+
+    /* Ensure little-endian byte order */
+    store32(outlen_bytes, (uint32_t)outlen);
+
+    if ((mac = EVP_MAC_fetch(NULL, "blake2bmac", NULL)) == NULL)
+        goto fail;
+
+    if ((ctx = EVP_MAC_new_ctx(mac)) == NULL)
+        goto fail;
+
+    par[par_n++] = OSSL_PARAM_construct_size_t(OSSL_MAC_PARAM_SIZE, &blake_outlen);
+    par[par_n++] = OSSL_PARAM_construct_end();
+
+    if (!EVP_MAC_set_ctx_params(ctx, par))
+        goto fail;
+
+    if (EVP_MAC_init(ctx) != 1)
+        goto fail;
+
+    if (EVP_MAC_update(ctx, outlen_bytes, sizeof(outlen_bytes)) != 1)
+        goto fail;
+
+    if (EVP_MAC_update(ctx, in, inlen) != 1)
+        goto fail;
+
+    if (outlen <= BLAKE2B_OUTBYTES) {
+        if (EVP_MAC_final(ctx, out, (size_t *) &out_written, outlen) != 1)
+            goto fail;
+    } else {
+        if (EVP_MAC_final(ctx, outbuf, (size_t *) &out_written,
+                          BLAKE2B_OUTBYTES) != 1)
+            goto fail;
+
+        memcpy(out, outbuf, BLAKE2B_OUTBYTES / 2);
+        out += BLAKE2B_OUTBYTES / 2;
+        outlen_curr = (uint32_t) outlen - BLAKE2B_OUTBYTES / 2;
+
+        while(outlen_curr > BLAKE2B_OUTBYTES) {
+            memcpy(inbuf, outbuf, BLAKE2B_OUTBYTES);
+            if (blake2b(outbuf, BLAKE2B_OUTBYTES, inbuf, BLAKE2B_OUTBYTES, NULL,
+                        0) != 1)
+                goto fail;
+            memcpy(out, outbuf, BLAKE2B_OUTBYTES / 2);
+            out += BLAKE2B_OUTBYTES / 2;
+            outlen_curr -= BLAKE2B_OUTBYTES / 2;
+        }
+
+        memcpy(inbuf, outbuf, BLAKE2B_OUTBYTES);
+        if (blake2b(outbuf, outlen_curr, inbuf, BLAKE2B_OUTBYTES, NULL, 0) != 1)
+            goto fail;
+        memcpy(out, outbuf, outlen_curr);
+    }
+    ret = 1;
+
+fail:
+    EVP_MAC_free_ctx(ctx);
+    EVP_MAC_free(mac);
+    return ret;
+}
+
+static void kdf_argon2_init(KDF_ARGON2 *ctx, ARGON2_TYPE type)
+{
+    ARGON2_CTX *c;
+
+    c = &ctx->ctx;
+    memset(c, 0, sizeof(*c));
+    c->outlen = 64;
```

Comment:
These magic numbers should be #defined (with the tag DEFAULT in the name)

Reply:
Defined.